[Feature] 많은 사람들이 관심을 가진 숙소 추천 API 구현 및 테스트완료

JJae-story opened the pull request:
## 📌 관련 이슈
- close #157 

## 📝 변경 사항
### AS-IS
- 로그인 유저 홈 화면에 조회수 기반 인기 숙소 추천 기능이 없어, 화면 구성이 부족하고 사용자 맞춤 콘텐츠가 부족했음.

### TO-BE
- 조회수 기준 상위 10개의 숙소를 추천하는 기능 추가
- 숙소는 `viewCount` 내림차순 + `totalRating` 내림차순으로 정렬하여 신뢰도 높은 숙소 우선 노출
- `AccommodationRepository`에 `findTop10ByOrderByViewCountDescTotalRatingDesc()` 메서드 추가
- 추천 결과는 숙소 id, 숙소명, 썸네일, 가격(최저가), 평점 정보를 포함하여 `RecommendationResponse`로 반환

## 🔍 테스트
- [x] 테스트 코드 작성
- [x] API 테스트
- [x] 로컬 테스트

## 📸 스크린샷
- id: 7은 조회수 4, id: 4, 8, 11은 같은 조회수 3
![많은 관심을 가진 숙소 추천 성공](https://github.com/user-attachments/assets/a98b5636-79df-40ca-9d2b-cae501bca46c)


## ✅ 체크리스트
- [x] main/develop 브랜치가 아닌 feature 브랜치에서 작성했나요?
- [x] 코딩 컨벤션을 준수했나요?
- [x] 불필요한 코드는 제거했나요?
- [x] 주석은 충분히 작성했나요?
- [x] 테스트는 완료했나요?

## 🙋‍♂️ 리뷰어에게
erd에 totalRating, viewCount 추가해놨습니다.
리뷰 부탁드립니다.

JJae-story committed:
feat: 엔티티에 viewCount 필드 추가

- erd에도 추가 완료

JJae-story committed:
feat: controller에 조회수 기반 인기 숙소 추천 추가

JJae-story committed:
feat: securityConfig에 유저만 접근할 수 있도록 설정 추가

JJae-story committed:
feat: 숙소 등록 dto에 viewCount 0으로 설정되도록 추가

JJae-story committed:
feat: 매달 1일에 숙소 조회수 초기화 하도록 추가

- scheduler 패키지를 만들어서 기존에 만들어둔 scheduler 도 옮김

JJae-story committed:
feat: service에 조회수 기반 인기 숙소 추천 로직 작성

JJae-story committed:
feat: 숙소 상세 조회 시, 해당 숙소 조회 수가 올라가도록 코드 추가.

- transactional(readOnly = true) 삭제

JJae-story committed:
feat: 테스트 코드 작성

JJae-story committed:
refactor: 조회수가 같은 경우 총 평점 높은 순 조건 추가

JJae-story committed:
refactor: url 변경, 주석 변경, 테스트코드 수정

dbghwns123 reviewed (commented):
첫 번째 커밋 메시지에 무슨 엔티티인지 남겨주시면 좋을 것 같습니다! 그 외의 커밋 메시지에서도 어떤 서비스 코드인지 남겨주시면 더 보기 편할 것 같습니다!

Seokjin-K commented on src/main/java/com/meongnyangerang/meongnyangerang/domain/accommodation/Accommodation.java:
```diff
@@ -64,7 +64,10 @@ public class Accommodation {
+  private Double totalRating;
+
+  @Column(nullable = false)
+  private Long viewCount;
```
viewCount 업데이트 과정에서 동시성 문제가 일어날 가능성이 있어 보이긴 하는데, 어떻게 생각하시나요?
저희 서비스가 얼마나 많은 트래픽을 받을 것인가 상정하는 게 포인트일 것 같고, viewCount의 정합성이 얼마나 높아야 하는가에 대해서도 고민할 부분인 것 같아요.
제 생각에는 트래픽이 얼마 없다면 정합성이 그렇게 중요한가 싶기는 해요.

JJae-story replied:
말씀하신 대로 현재 방식에서는 락을 걸고 사용하지 않아서, 동시성 문제가 발생할 수 있을 것 같긴 하네요.
제가 생각하기에도 저희 서비스가 많은 트래픽이 발생할 것 같지는 않은데, 정합성은 중요하지 않을까요?
현재 방식 (db에서 숙소 객체 가져오기 -> set -> save) 에서 viewCount 증가 시키는 방식을 DB에서 쿼리로 증가시키면  동시성 문제는 발생하지 않을 것 같은데, 어떻게 생각하시나요?

dbghwns123 replied:
제 생각에도 현재 방식(객체를 가져와서 set으로 값을 변경하고 다시 저장)은 낙관적 락이나 비관적 락을 사용하지 않는 한, 동시성 문제가 발생할 가능성이 충분히 있다고 생각됩니다. 의재님과 석진님이 말씀해주신 것처럼 저희 서비스가 얼마나 많은 트래픽이 발생할지가 키 포인트일 것 같네요. 이 부분 해결하기 위해서 DB에서 쿼리로 viewCount를 증가시키는 방식은 동시성 문제를 해결하는 데 효과적일 것 같습니다!

Seokjin-K replied:
현재 DB 쿼리로 증가시키고 있나요? dirty checking으로 증가되는 거 아닌가요?
+
viewCount 증가 부분은 비동기로 처리해도 좋을 것 같은데 어떻게 생각하시나요?

JJae-story replied:
아뇨, 현재는 dirty checking으로 증가되는 게 아니라 save() 메서드가 호출되며 명시적으로 쿼리를 발생시키고 있습니다. -> 아 숙소를 조회하는 시점에서 영속성 컨텍스트에서 관리 되니 dirty checking 이 작동하는 건 맞네요.
비동기로 처리하는 건 안해봐서 잘 모르는데, 동시성 문제를 어떻게 해결할 수 있나요? 또, DB 에서 쿼리로 증가시키는 방식보다 안전할까요?

dbghwns123 replied:
저도 비동기에 대해서 잘 알지 못해서 찾아본 결과, 제 생각에 지금 이 상황에서는 의재님이 말씀하신 viewCount 증가를 DB에서 쿼리로 증가시켜서 동시성 문제를 해결하는게 효율적이지 않을까 생각합니다. 비동기 처리는 요청을 병렬로 실행하지만, viewCount 증가의 원자성을 보장하지 않을 뿐더러 비동기로 viewCount를 증가시켜도, 동일한 엔티티를 동시에 수정하면 문제가 발생할 것이라고 생각되는데 어떻게 생각하시나요?

Seokjin-K replied:
아 동시성 처리를 위해 비동기로 처리한다기 보단, 동시성 처리를 락으로 처리하실 거라면 viewCount 올리는 부분만 비동기로 처리하면 락으로 인한 병목현상이 줄어들 것 같아서 말씀드렸습니다.

Seokjin-K reviewed (commented):
수고하셨습니다!

Seokjin-K commented on src/main/java/com/meongnyangerang/meongnyangerang/service/AccommodationService.java:
```diff
@@ -176,10 +175,17 @@ public AccommodationDetailResponse getAccommodationDetail(Long accommodationId)
     List<Review> reviews = reviewRepository.findTop5ByAccommodationIdOrderByCreatedAtDesc(
         accommodationId);
 
+    updateViewCount(accommodation);
```
@Transactional 없이 viewCount가 잘 증가되나요?

JJae-story replied:
네, 해당 부분은 테스트 해서 확인하였습니다.

dbghwns123 commented on src/main/java/com/meongnyangerang/meongnyangerang/service/AccommodationService.java:
```diff
@@ -176,10 +175,17 @@ public AccommodationDetailResponse getAccommodationDetail(Long accommodationId)
   }
 
+  public void updateViewCount(Accommodation accommodation) {
+    accommodation.setViewCount(accommodation.getViewCount() + 1);
```
이 부분에서 setter를 사용하여 객체의 상태를 직접 변경하고 계시는데 이 방식보다는 `Accommodation` 엔티티에 `incrementViewCount()`와 같은 메서드를 추가하여 뷰 카운트를 증가시키는 로직을 캡슐화하는 것이 객체지향 설계 원칙 측면(객체가 자신의 상태를 관리하도록 책임을 부여하는 것)에서 봤을때 더 적절해 보이는데 어떻게 생각하시나요?

JJae-story replied:
넵, 현재 방식이라면 그렇게 변경하는게 더 좋아보이긴 합니다. 위에 석진님이 동시성 문제를 말씀해 주셔서 답글을 달아놨는데, 한번 확인 부탁드리고 의견 부탁드립니다!

dbghwns123 commented on src/main/java/com/meongnyangerang/meongnyangerang/service/AccommodationService.java:
```diff
@@ -149,7 +149,6 @@ public AccommodationResponse updateAccommodation(
   /**
    * 숙소 상세 조회(비로그인 사용자, 일반 사용자, 호스트 모두 접근 가능한 API)
    */
-  @Transactional(readOnly = true)
```
기존에는 숙소 상세 조회 기능이여서 @Transactional(readOnly = true)을 추가했었던 부분인데 지금은 아예 삭제하신걸로 보이네요. @Transactional 이 있어야 하지 않을까요?

JJae-story replied:
기존의 @Transactional(readOnly = true)를 제거한 이유는 viewCount 업데이트가 필요했기 때문입니다.
readOnly = true가 설정되어 있으면 내부적으로 flush가 발생하지 않아 update 쿼리가 실행되지 않기 때문에 제거했습니다.
원래는 GET 요청이라 트랜잭션 자체를 생략했었는데, 쿼리 방식으로 업데이트를 처리하게 되면 일반 트랜잭션을 추가해야 할 것 같습니다.

dbghwns123 replied:
아 제 말은 제가 기존에 숙소 상세 조회 기능을 구현할때는 조회 기능만 있어서 @transactional(readOnly = true)을 추가했었는데 지금 의재님이 추가하신 부분에 있어서는 당연히 (readOnly = true)는 빠져야 한다고 생각됩니다. 제가 말씀드린건 updateViewCount 메서드에서 viewCount를 하나 증가한 후 save를 하는과정에서는 @Transactional이 필요하지 않을까 해서 말씀드렸습니다!

JJae-story replied:
아 현재 방식에서는 save()를 호출해주고 있기 때문에, 저희가 사용하고 있는 JpaRepository가 구현체 클래스인SimpleJpaRepository를 참조하고 있으며, 이 구현체 클래스는 내부적으로 수정, 삭제 등에 트랜잭션을 가지고 있어서 명시적으로 안붙여놨었습니다.
쿼리 방식으로 바꾸게 되면 붙이게 될꺼 같습니다!

dbghwns123 replied:
아 그렇군요 SimpleJpaRepository를 참조하고 이 클래스는 내부적으로 트랜잭션을 가지고 있는 부분은 알지 못했는데 새로운 걸 알아가네요 감사합니다!

JJae-story committed:
fix: 숙소 조회 시 뷰 카운트 증가 쿼리 수정

- 숙소 상세 조회 시 뷰 카운트를 증가시키는 로직을 @Modifying 어노테이션을 사용한 쿼리로 수정
- `incrementViewCount` 메서드를 레포지토리에 추가하여 직접 업데이트 쿼리를 실행하도록 구현
- 서비스 계층에서 해당 쿼리를 호출하도록 수정
- 트랜잭션을 명시적으로 적용하여 뷰 카운트 증가와 관련된 동시성 문제 해결

dbghwns123 reviewed (approved):
수고하셨습니다!

Seokjin-K reviewed (approved):
수고하셨습니다!
